### PR TITLE
Add @process endpoint supporting nested task sequences.

### PR DESCRIPTION
**Pull request by njohner**
In this PR we implement a new endpoint `@process` which allows to create task sequences. This endpoint is a substitute for the `@trigger-task-template`, but allowing more flexibility. It notably supports nested task sequences and the created process does not need to come from a `TaskTemplateFolder`. The old functionality and endpoint is kept for now, so that we do not have an API change. A few notable points:
- The `@process` can now validate the data of subtasks directly using the `ITask` schema, as we do not have half the data coming from the corresponding `TaskTemplate` which forced us to validate everything by hand before.
- All fields from `ITask` are now supported for the subtasks
- Validation, deadline calculation, data preprocessing and process creation are now all recursive and support any nesting depth

A few things that still need to be addressed:
- The interfaces used for marking Sequential and Parallel processes (`IFromSequentialTasktemplate`, `IFromParallelTasktemplate`) are not sufficient anymore as is, because we were relying on the main task and the subtasks to provide the same interface. With nesting this is not possible anymore, because a sequential `TasktemplateFolder` can be nested into a parallel one, so that the created task should have both interfaces, `IFromParallelTasktemplate` because it is part of a parallel process because of its parent, and `IFromSequentialTasktemplate` because it is the parent of a sequential set of tasks. This leads to various issues and needs to be looked at, notably:
  - Initial states of the created Tasks are not yet as expected (This is at least partially linked to the next point)
  - tasktemplate predecessor is not set correctly
- Check start next
- Check event handlers
- Documentation

Definition of Done: https://4teamwork.atlassian.net/wiki/spaces/CHX4TW/pages/917562/

For [CA-3725]

## Checklist
- [ ] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)
- API change:
  - [ ] Documentation is updated
  - [ ] API Changelog entry (see [guide](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/451248812/API+Changelog+Guidelines))
  - If breaking:
    - [ ] api-change label added
    - [ ] #delivery channel notified about breaking change
    - [ ] Scrum master is informed
- Further improvements needed:
  - [ ] Create follow-up stories and link them in the PR and Jira issue

[CA-3725]: https://4teamwork.atlassian.net/browse/CA-3725?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: opengever/tasktemplates/tasktemplatefolder.py [hunk 1]
```diff
@@ -182,8 +175,25 @@ def preprocess_data(self):
 
     def _recursive_preprocess_data(self, data):
         self.replace_interactive_actors(data)
-        for item in data.get("items", []):
-            self._recursive_preprocess_data(item)
+        if self.has_children(data):
```
Comment (phgross) [inline opengever/tasktemplates/tasktemplatefolder.py:166]: As discussed, I'm not sure this calculation makes sense here. But I think it is ok for now and can be adjusted in a separate PR.

File: opengever/tasktemplates/tasktemplatefolder.py [hunk 2]
```diff
@@ -238,14 +238,27 @@ def create_main_task(self, data):
...
         return main_task
 
+    @staticmethod
+    def set_state(task, state):
+        initial_state = api.content.get_state(task)
+        wftool = api.portal.get_tool('portal_workflow')
+        wf_id = wftool.getWorkflowsFor(task)[0].id
+        wftool.setStatusOf(wf_id, task, {'review_state': state})
+        wftool.getWorkflowsFor(task)[0].updateRoleMappingsFor(task)
+        return initial_state
```
Comment (phgross) [inline opengever/tasktemplates/tasktemplatefolder.py:258]: When changing the `state` like this, its not guaranteed that the representation in the globalindex is updated. Maybe extend the tests to make it sure.

File: opengever/api/process.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    @staticmethod
+    def is_task_container(data):
+        if "items" in data or "sequence_type" in data:
+            return True
+        return False
+
+    @staticmethod
+    def raise_bad_request(errors):
+        errors = [{"field": each[0], "error": each[1]} for each in errors]
+        raise BadRequest(errors)
```
Comment (phgross) [inline opengever/api/process.py:188]: Are the errors already translated and can be shown correctly in the UI?

**Review by phgross — APPROVED** [review]
Nice work 👍 

Only a few question marks and comments.

File: opengever/tasktemplates/tasktemplatefolder.py (hunk 1 above)
Comment (njohner) [reply opengever/tasktemplates/tasktemplatefolder.py:166]: Yes I agree. I think we would still need the recursive calculation, but maybe not add the 5 days at the end, so would be an easy fix

File: opengever/tasktemplates/tasktemplatefolder.py (hunk 2 above)
Comment (njohner) [reply opengever/tasktemplates/tasktemplatefolder.py:258]: We don't really want the global index to be updated, as this is a temporary change. In create_subtasks we change the state to task-state-in-progress but a few lines after we set the state back to its initial state.
I'll add a test none the less, as there is a risk that this suddenly gets used differently in the future.

File: opengever/api/process.py (hunk 3 above)
Comment (njohner) [reply opengever/api/process.py:188]: Formatting the errors as I do here allows them to be recognised as validation errors in our error handling and therefore be "translated" (basically just the "invalid input" is translated). Frontend will have a hard time showing the errors correctly though, as in the recursive schema, each field appears multiple times, with no real way of distinguishing them. We would need to invest an unreasonable amount of time here IMO if we want to do something better, knowing that the frontend does not handle any form validations correctly currently (it only has its own form validation but does not handle validation errors returned by the API).

File: opengever/api/tests/test_process.py [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+            }
+        }
+
+        with browser.expect_http_error(400):
+            browser.open('{}/@process'.format(
+                         self.dossier.absolute_url()),
+                         data=json.dumps(data),
+                         headers=self.api_headers)
+
+        self.assertEqual(u'Inputs not valid',
```
Comment (njohner) [inline opengever/api/tests/test_process.py:380]: Here you see that the error is translated

File: opengever/api/tests/test_process.py [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+                         browser.json[u'translated_message'])
+
+        self.assertItemsEqual(
+            [{u'field': u'responsible_client',
+              u'translated_message': u'None',
+              u'type': u'responsible_client'},
+             {u'field': u'responsible',
+              u'translated_message': u'None',
+              u'type': u'responsible'}],
+            browser.json[u'additional_metadata']['fields'])
```
Comment (njohner) [inline opengever/api/tests/test_process.py:390]: with `additional_metadata` that is only somewhat helpful
